audit-base/src/main/java/com/wultra/core/audit/base/database/DatabaseAuditWriter.java

```diff
@@ -0,0 +1,244 @@
+                                        ps.setString(5, throwable.getMessage());
+                                        ps.setString(6, sw.toString());
+                                    }
+                                    ps.setString(7, jsonUtil.serializeMap(record.getParam()));
+                                    ps.setString(8, record.getCallingClass().getName());
+                                    ps.setString(9, record.getThreadName());
+                                    ps.setString(10, version);
+                                    ps.setTimestamp(11, new Timestamp(buildTime.toEpochMilli()));
+                                    for (int j = 0; j < paramColumnNames.size(); j++) {
+                                        int index = 12 + j;
+                                        String paramName = paramColumnNames.get(j).replace("param_", "");
+                                        Object paramValue = record.getParam().get(paramName);
+                                        if (paramValue == null) {
+                                            ps.setNull(index, Types.VARCHAR);
+                                        } else {
+                                            if (paramValue instanceof CharSequence) {
+                                                ps.setString(index, paramValue.toString());
+                                            } else {
+                                                ps.setString(index, jsonUtil.serializeObject(paramValue));
+                                            }
+                                        }
+                                    }
+                                }
+
+                                public int getBatchSize() {
+                                    return recordsToPersist.size();
+                                }
+                            });
+                    logger.debug("Audit log batch insert succeeded, record count: {}", insertCounts.length);
+                } catch (InterruptedException ex) {
+                    logger.warn(ex.getMessage(), ex);
+                }
+            }
+        }
+    }
+
+    public void cleanup() {
+        if (jdbcTemplate.getDataSource() == null) {
+            logger.error("Data source is not available");
+            return;
+        }
+        LocalDateTime cleanupLimit = LocalDateTime.now().minusDays(cleanupDays);
+        synchronized (CLEANUP_LOCK) {
+            jdbcTemplate.execute( SQL_DELETE_FROM + tableName + SQL_CLEANUP_CONDITION, (PreparedStatementCallback<Boolean>) ps -> {
+                ps.setTimestamp(1, Timestamp.valueOf(cleanupLimit));
+                return ps.execute();
+            });
+            logger.debug("Audit records older than {} were deleted", cleanupLimit);
+        }
+    }
+
+    @Scheduled(fixedDelayString = "${audit.flush.delay.fixed:1000}", initialDelayString = "${powerauth.audit.flush.delay.initial:1000}")
```

Comment:
This will cause issues when running in multiple instances. I suggest either:

- Removing scheduled cleanup from the library and delegating this to the application (including documentation).
- Implementing some multi-node locking (Shedlock with JDBC).

Reply:
There is no issue with created audit records, each instance creates is own records, so no locking is required. 

During cleanup the issue is also not as significant as it seems, because:
- multiple `DELETE` queries can run in parallel, no locking is required, the DB can handle this easily
- the `DELETE` queries are by default executed once per hour by default (this value can be customized) and are very likely to be executed in different times by different nodes
- an index is created on the `created_timestamp` column, so the `DELETE` queries should be fast

We can add a locking solution, but I think it is an overkill for this use case.